core/server/master/src/main/java/alluxio/master/file/DefaultFileSystemMaster.java

```diff
@@ -1458,40 +1459,51 @@ public boolean exists(AlluxioURI path, ExistsContext context)
           DescendantType.ONE, auditContext, LockedInodePath::getInodeOrNull,
           false);
 
-      try (LockedInodePath inodePath = mInodeTree.lockInodePath(
-          createLockingScheme(path, context.getOptions().getCommonOptions(), LockPattern.READ),
-          rpcContext.getJournalContext())
-      ) {
-        LoadMetadataContext lmCtx = LoadMetadataContext.create(
-            LoadMetadataPOptions.newBuilder()
-                .setCommonOptions(context.getOptions().getCommonOptions())
-                .setLoadType(context.getOptions().getLoadMetadataType()));
-        if (shouldLoadMetadataIfNotExists(inodePath, lmCtx)) {
-          checkLoadMetadataOptions(context.getOptions().getLoadMetadataType(), path);
+      LoadMetadataContext lmCtx = LoadMetadataContext.create(
+          LoadMetadataPOptions.newBuilder()
+              .setCommonOptions(context.getOptions().getCommonOptions())
+              .setLoadType(context.getOptions().getLoadMetadataType()));
+      /*
+      See the comments in #getFileIdInternal for an explanation on why the loop here is required.
+       */
+      boolean run = true;
+      boolean loadMetadata = false;
+      while (run) {
+        run = false;
+        if (loadMetadata) {
+          try {
+            checkLoadMetadataOptions(context.getOptions().getLoadMetadataType(), path);
+          } catch (FileDoesNotExistException e) {
+            return false;
+          }
           loadMetadataIfNotExist(rpcContext, path, lmCtx, false);
         }
-      } catch (FileDoesNotExistException e) {
-        // ignore exception
-      }
 
-      try (LockedInodePath inodePath = mInodeTree.lockInodePath(
-          createLockingScheme(path, context.getOptions().getCommonOptions(), LockPattern.READ),
-          rpcContext.getJournalContext())
-      ) {
-        try {
-          if (mPermissionChecker instanceof DefaultPermissionChecker) {
-            mPermissionChecker.checkParentPermission(Mode.Bits.EXECUTE, inodePath);
+        try (LockedInodePath inodePath = mInodeTree.lockInodePath(
+            createLockingScheme(path, context.getOptions().getCommonOptions(), LockPattern.READ),
+            rpcContext.getJournalContext())
+        ) {
+          if (!loadMetadata && shouldLoadMetadataIfNotExists(inodePath, lmCtx)) {
+            loadMetadata = true;
+            run = true;
+            continue;
           }
-        } catch (AccessControlException e) {
-          auditContext.setAllowed(false);
-          throw e;
+          try {
```

Comment:
see the code you referred to in the link, the perm checker happened in a different order. Should you do the same?

Reply:
the existing exists() method loads the metadata before it actually checks the permission. I don't want to introduce a semantics change here.